src/bvh/bvh_node.rs

```diff
@@ -0,0 +1,608 @@
+                let shape_aabb = shape.aabb();
+                let shape_center = shape_aabb.center();
+
+                // Get the relative position of the shape centroid `[0.0..1.0]`.
+                let bucket_num_relative =
+                    (shape_center[split_axis] - centroid_bounds.min[split_axis]) / split_axis_size;
+
+                // Convert that to the actual `Bucket` number.
+                let bucket_num = (bucket_num_relative
+                    * (T::from(NUM_BUCKETS).unwrap() - T::from(0.01).unwrap()))
+                .to_usize()
+                .unwrap();
+
+                // Extend the selected `Bucket` and add the index to the actual bucket.
+                buckets[bucket_num].add_aabb(&shape_aabb);
+                bucket_assignments[bucket_num].push(*idx);
+            }
+
+            // Compute the costs for each configuration and select the best configuration.
+            let mut min_bucket = 0;
+            let mut min_cost = T::infinity();
+            let mut child_l_aabb = Aabb::empty();
+            let mut child_l_centroid = Aabb::empty();
+            let mut child_r_aabb = Aabb::empty();
+            let mut child_r_centroid = Aabb::empty();
+            for i in 0..(NUM_BUCKETS - 1) {
+                let (l_buckets, r_buckets) = buckets.split_at(i + 1);
+                let child_l = l_buckets.iter().fold(Bucket::empty(), Bucket::join_bucket);
+                let child_r = r_buckets.iter().fold(Bucket::empty(), Bucket::join_bucket);
+
+                let cost = (T::from(child_l.size).unwrap() * child_l.aabb.surface_area()
+                    + T::from(child_r.size).unwrap() * child_r.aabb.surface_area())
+                    / aabb_bounds.surface_area();
+                if cost < min_cost {
+                    min_bucket = i;
+                    min_cost = cost;
+                    child_l_aabb = child_l.aabb;
+                    child_l_centroid = child_l.centroid;
+                    child_r_aabb = child_r.aabb;
+                    child_r_centroid = child_r.centroid;
+                }
+            }
+            // Join together all index buckets.
+            // split input indices, loop over assignments and assign
+            let (l_assignments, r_assignments) = bucket_assignments.split_at_mut(min_bucket + 1);
+
+            let mut l_count = 0;
+            for group in l_assignments.iter() {
+                l_count += group.len();
+            }
+
+            let (child_l_indices, child_r_indices) = indices.split_at_mut(l_count);
+            let mut i = 0;
+            for group in l_assignments.iter() {
+                for x in group {
+                    child_l_indices[i] = *x;
+                    i += 1;
+                }
+            }
+            i = 0;
+            for group in r_assignments.iter() {
+                for x in group {
+                    child_r_indices[i] = *x;
+                    i += 1;
+                }
+            }
```

Comment:
Please use separate indices for these. Maybe `l_i` and `r_i`? That will decrease the chance of fuckups when this code is refactored sometime down the road.

Reply:
sgtm, I just slightly refactored both of these loops to use iterators more